From 9db9166b7c97dda34167a06b65b627ca4dbf2052 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:07:04 -0400
Subject: [powerpc] vio: Fix a dma_mask issue of vio

Message-id: <1410545655-205645-196-git-send-email-dzickus@redhat.com>
Patchwork-id: 94672
O-Subject: [RHEL7 PATCH 195/626] powerpc/vio: Fix a dma_mask issue of vio
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit c6102609289de268bcc6093e606483ea8bb35c84
Author: Li Zhong <zhong@linux.vnet.ibm.com>
Date:   Tue Nov 19 16:11:37 2013 +0800

    powerpc/vio: Fix a dma_mask issue of vio

    I encountered following issue:
    [    0.283035] ibmvscsi 30000015: couldn't initialize event pool
    [    5.688822] ibmvscsi: probe of 30000015 failed with error -1

    which prevents the storage from being recognized, and the machine from
    booting.

    After some digging, it seems that it is caused by commit 4886c399da

    as dma_mask pointer in viodev->dev is not set, so in
    dma_set_mask_and_coherent(), dma_set_coherent_mask() is not called
    because dma_set_mask(), which is dma_set_mask_pSeriesLP() returned EIO.
    While before the commit, dma_set_coherent_mask() is always called.

    I tried to replace dma_set_mask_and_coherent() with
    dma_coerce_mask_and_coherent(), and the machine could boot again.

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/vio.c b/arch/powerpc/kernel/vio.c
index cb92d82..aba1540 100644
--- a/arch/powerpc/kernel/vio.c
+++ b/arch/powerpc/kernel/vio.c
@@ -1413,8 +1413,7 @@ struct vio_dev *vio_register_device_node(struct device_node *of_node)
 
   /* needed to ensure proper operation of coherent allocations
    * later, in case driver doesn't set it explicitly */
-  dma_set_mask(&viodev->dev, DMA_BIT_MASK(64));
-  dma_set_coherent_mask(&viodev->dev, DMA_BIT_MASK(64));
+  dma_coerce_mask_and_coherent(&viodev->dev, DMA_BIT_MASK(64));
  }
 
  /* register with generic device framework */
-- 
1.7.1